src/crimson/os/seastore/seastore_types.h

```diff
@@ -1051,34 +1071,172 @@ class laddr_t {
 
   friend auto operator<=>(const laddr_t&, const laddr_t&) = default;
 
-  friend laddr_t operator+(const laddr_t &laddr, const Unsigned &i) {
-    return laddr_t{laddr.value + i};
-  }
-
-  friend laddr_t operator+(const Unsigned &i, const laddr_t &laddr) {
-    return laddr_t{laddr.value + i};
-  }
-
-  friend laddr_t operator-(const laddr_t &laddr, const Unsigned &i) {
-    return laddr_t{laddr.value - i};
-  }
-
+  // the difference between two laddrs usually works as byte distance,
+  // assume the result won't overflow and convert it intto the byte
+  // addressing result automatically.
   friend Unsigned operator-(const laddr_t &l, const laddr_t &r) {
-    return l.value - r.value;
+    auto diff = l.value - r.value;
+    assert(diff <= (std::numeric_limits<Unsigned>::max() >> UNIT_SHIFT));
+    return diff << UNIT_SHIFT;
   }
 
+  friend struct laddr_offset_t;
   friend struct laddr_le_t;
   friend struct pladdr_le_t;
 
 private:
+  // Prevent direct construction of laddr_t with an integer,
+  // always use laddr_t::from_raw_uint instead.
+  constexpr explicit laddr_t(Unsigned value) : value(value) {}
   Unsigned value;
 };
 
-constexpr laddr_t L_ADDR_MAX = laddr_t(laddr_t::RAW_VALUE_MAX);
-constexpr laddr_t L_ADDR_MIN = laddr_t(0);
+constexpr laddr_t L_ADDR_MAX = laddr_t::from_raw_uint(laddr_t::RAW_VALUE_MAX);
+constexpr laddr_t L_ADDR_MIN = laddr_t::from_raw_uint(0);
 constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
-constexpr laddr_t L_ADDR_ROOT = laddr_t(laddr_t::RAW_VALUE_MAX - 1);
-constexpr laddr_t L_ADDR_LBAT = laddr_t(laddr_t::RAW_VALUE_MAX - 2);
+constexpr laddr_t L_ADDR_ROOT = laddr_t::from_raw_uint(laddr_t::RAW_VALUE_MAX - 1);
+constexpr laddr_t L_ADDR_LBAT = laddr_t::from_raw_uint(laddr_t::RAW_VALUE_MAX - 2);
+
+// laddr_offset_t contains one base laddr and one block not aligned
+// offset(< laddr_t::UNIT_SIZE). It's the return type of plus/subtract
+// overloads for laddr_t and loffset_t.
+//
+// NOTE: Always use check_to_laddr() to get alinged laddr_t when we could
+// ensure the offset is alinged, e.g. in LBAManager.
+struct laddr_offset_t {
+  explicit laddr_offset_t(laddr_t base)
+      : base(base), offset(0) {}
+  laddr_offset_t(laddr_t base, loffset_t offset)
+      : base(base), offset(offset) {}
+
+  laddr_t get_roundup_laddr() const {
+    if (offset == 0) {
+      return base;
+    } else {
+      assert(offset < laddr_t::UNIT_SIZE);
+      return laddr_t(base.value + 1);
+    }
+   }
+  laddr_t get_aligned_laddr() const { return base; }
+  loffset_t get_offset() const { return offset; }
```

Comment:
Let change the type of `laddr_t::offset` to `extent_len_t` and return `extent_len_t` here.

This is logically correct because `UNIT_SIZE` should be smaller than `extent_len_t`, and `extent_len_t` smaller/eq than `objaddr_t`, and `objaddr_t` + `laddr_t::get_offset()` won't be promoted to `loffset_t`.